File: pixel.py
from math import *

class Pixel:
	def __init__(self, r, g, b):
		self.r = int(r)
		self.g = int(g)
		self.b = int(b)

	# add RGB values to pixel buffer
	def add(self, buffer):
		buffer.append(self.r)
		buffer.append(self.g)
		buffer.append(self.b)

MAX  = 1000
BAIL = 4.0

# find RGB values for pixel at (px, py)
def calc(px, py):
	
	# find number of iterations	to escape
	x = 0.0
	y = 0.0
	it = 0
	
	# iterations to escape
	while (x**2 + y**2 <= BAIL and it < MAX):
		tmp = x**2 - y**2 + px
		y = 2*x*y + py
		x = tmp
		it += 1

	if (it == MAX):
		return color(MAX)

	if (it + 1 == MAX):
		return color(it)
	
	# smooth coloring with interpolation
	zn = log(x**2 + y**2) / 2
	nu = log(zn / log(2)) / log(2)
	it = it + 1 - nu

	# end points
	a = color(floor(it))
	b = color(floor(it) + 1)

	# final color
	r = interpolate(a.r, b.r, it % 1)
	g = interpolate(a.g, b.g, it % 1)
	b = interpolate(a.b, b.b, it % 1)

	return Pixel(r, g, b)

# return RGB given value in range [0, MAX]
def color(value):
	
	# normalize value
	normal = value / MAX

	r = normal * 35
	g = normal * 35
	b = normal * 200

	return Pixel(r, g, b)

# linear interpolation
def interpolate(a, b, t):
	u = 1 - t
	return a*u + b*t

File: gui/pixel.cpp
#include "pixel.h"
#include <cmath>

rgb_t buf[WIDTH][HEIGHT];

void calc_pixels()
{
	/* calculate new RGB values for each pixel */
	for (int px = 0; px < WIDTH; ++px) {
		for (int py = 0; py < HEIGHT; ++py) {
			buf[px][py] = calc_pixel(px, py);
		}
	}
}

rgb_t calc_pixel(double px, double py)
{
	/* translate pixel coordinates to actual coordinates */
	double ax = state.ax(px);
	double ay = state.ay(py);

	double x = 0;
	double y = 0;
	int i = 0;	

	/* escape time algorithm */
	while (pow(x, 2) + pow(y, 2) < THRESHOLD && i < MAX_ITER) {
		double xtemp = pow(x, 2) - pow(y, 2) + ax;
		y = 2*x*y + ay;
		x = xtemp;
		++i;
	}

	/* color inside pixels black */
	if (i == MAX_ITER) {
		return { 0, 0, 0 };
	}

	/* no interpolation for last bucket */
	if (i + 1 == MAX_ITER) {
		return color(i);
	}

	/* smooth coloring algorithm */
	double zn = log(pow(x, 2) + pow(y, 2)) / 2;
	double nu = log(zn / log(2)) / log(2);
	double it = i + 1 - nu;

	/* end points */
	rgb_t a = color(floor(it));
	rgb_t b = color(floor(it) + 1);

	/* interpolated color */
	return {
		linear(a.R, b.R, fmod(it, 1)),
		linear(a.G, b.G, fmod(it, 1)),
		linear(a.B, b.B, fmod(it, 1))
	};
}

rgb_t color(double a)
{
	return { a * 35, a * 35, a * 200 };
}

double linear(double a, double b, double t)
{
	double u = 1 - t;
	return a*u + b*t;
}
